File: libshvcore/src/chainpack/rpcdriver.cpp
#include "rpcdriver.h"
#include "metatypes.h"
#include "chainpackprotocol.h"
#include "../core/shvexception.h"
#include "../core/log.h"

#include <sstream>
#include <iostream>

#define logRpc() shvCDebug("rpc")

namespace shv {
namespace core {
namespace chainpack {

int RpcDriver::s_defaultRpcTimeout = 5000;

RpcDriver::RpcDriver()
{
}

RpcDriver::~RpcDriver()
{
}

void RpcDriver::sendMessage(const shv::core::chainpack::RpcValue &msg)
{
	using namespace std;
	//shvLogFuncFrame() << msg.toStdString();
	std::ostringstream os_packed_data;
	shv::core::chainpack::ChainPackProtocol::write(os_packed_data, msg);
	std::string packed_data = os_packed_data.str();
	logRpc() << "send message: packed data: " << (packed_data.size() > 50? "<... long data ...>" : packed_data);

	enqueueDataToSend(Chunk{std::move(packed_data)});
}

void RpcDriver::enqueueDataToSend(RpcDriver::Chunk &&chunk_to_enqueue)
{
	/// LOCK_FOR_SEND lock mutex here in the multothreaded environment
	if(!chunk_to_enqueue.empty())
		m_chunkQueue.push_back(std::move(chunk_to_enqueue));
	if(!isOpen()) {
		shvError() << "write data error, socket is not open!";
		return;
	}
	flushNoBlock();
	writeQueue();
	/// UNLOCK_FOR_SEND unlock mutex here in the multothreaded environment
}

namespace {
const unsigned PROTOCOL_VERSION = 1;
}

void RpcDriver::writeQueue()
{
	if(m_chunkQueue.empty())
		return;
	logRpc() << "writePendingData(), HI prio queue len:" << m_chunkQueue.size();
	//static int hi_cnt = 0;
	const Chunk &chunk = m_chunkQueue[0];

	if(m_headChunkBytesWrittenSoFar == 0) {
		std::string protocol_version_data;
		{
			std::ostringstream os;
			shv::core::chainpack::ChainPackProtocol::writeUIntData(os, PROTOCOL_VERSION);
			protocol_version_data = os.str();
		}
		{
			std::ostringstream os;
			shv::core::chainpack::ChainPackProtocol::writeUIntData(os, chunk.length() + protocol_version_data.length());
			std::string packet_len_data = os.str();
			auto len = writeBytes(packet_len_data.data(), packet_len_data.length());
			if(len < 0)
				SHV_EXCEPTION("Write socket error!");
			if(len < (int)packet_len_data.length())
				SHV_EXCEPTION("Design error! Chunk length shall be always written at once to the socket");
		}
		{
			auto len = writeBytes(protocol_version_data.data(), protocol_version_data.length());
			if(len < 0)
				SHV_EXCEPTION("Write socket error!");
			if(len != 1)
				SHV_EXCEPTION("Design error! Protocol version shall be always written at once to the socket");
		}
	}

	{
		auto len = writeBytes(chunk.data() + m_headChunkBytesWrittenSoFar, chunk.length() - m_headChunkBytesWrittenSoFar);
		if(len < 0)
			SHV_EXCEPTION("Write socket error!");
		if(len == 0)
			SHV_EXCEPTION("Design error! At least 1 byte of data shall be always written to the socket");

		logRpc() << "writeQueue - data len:" << chunk.length() << "start index:" << m_headChunkBytesWrittenSoFar << "bytes written:" << len << "remaining:" << (chunk.length() - m_headChunkBytesWrittenSoFar - len);
		m_headChunkBytesWrittenSoFar += len;
		if(m_headChunkBytesWrittenSoFar == chunk.length()) {
			m_headChunkBytesWrittenSoFar = 0;
			m_chunkQueue.pop_front();
		}
	}
}

void RpcDriver::bytesRead(std::string &&bytes)
{
	logRpc() << bytes.length() << "bytes of data read";
	m_readData += std::string(std::move(bytes));
	while(true) {
		int len = processReadData(m_readData);
		//shvInfo() << len << "bytes of" << m_readData.size() << "processed";
		if(len > 0)
			m_readData = m_readData.substr(len);
		else
			break;
	}
}

int RpcDriver::processReadData(const std::string &read_data)
{
	logRpc() << __FUNCTION__ << "data len:" << read_data.length();
	using namespace shv::core::chainpack;

	std::istringstream in(read_data);

	uint64_t chunk_len = ChainPackProtocol::readUIntData(in);

	size_t read_len = (size_t)in.tellg() + chunk_len;

	uint64_t protocol_version = ChainPackProtocol::readUIntData(in);

	if(protocol_version != PROTOCOL_VERSION)
		SHV_EXCEPTION("Unsupported protocol version");

	logRpc() << "\t chunk len:" << chunk_len << "read_len:" << read_len << "stream pos:" << in.tellg();
	if(read_len > read_data.length())
		return 0;

	RpcValue msg = ChainPackProtocol::read(in);
	onMessageReceived(msg);
	return read_len;
}

void RpcDriver::onMessageReceived(const RpcValue &msg)
{
	logRpc() << "\t emitting message received:" << msg.toStdString();
	//logLongFiles() << "\t emitting message received:" << msg.dumpText();
	if(m_messageReceivedCallback)
		m_messageReceivedCallback(msg);
}

} // namespace chainpack
} // namespace core
} // namespace shv

File: python/chainpack/rpcdriver.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from value import *
import logging
from rpcmessage import *


logger=logging.getLogger()
logger.setLevel(logging.DEBUG)
from utils import print_to_string
def log(*vargs):
	logger.debug(print_to_string(*vargs))
def info(*vargs):
	logger.info(print_to_string(*vargs))


defaultRpcTimeout = 5000;

class RpcDriver():
	PROTOCOL_VERSION = 1;

	def __init__(s):
		s.m_readData = bytearray()

	def sendMessage(s, msg: RpcValue):
		chunk = ChainPackProtocol(msg)
		log("send message: packed data: ",  chunk[:50] + "<... long data ...>" if len(chunk) > 50 else chunk)
		protocol_version_data = ChainPackProtocol()
		protocol_version_data.writeData_UInt(s.PROTOCOL_VERSION)
		packet_len_data = ChainPackProtocol()
		packet_len_data.writeData_UInt(len(protocol_version_data) + len(chunk))
		s.writeBytes(packet_len_data)
		s.writeBytes(protocol_version_data)
		s.writeBytes(chunk)

	def bytesRead(s, b: bytes):
		if len(b) == 0:
			return
		log(len(b), "bytes of data read")
		s.m_readData += b
		while True:
			l: int = s.processReadData(s.m_readData);
			log(l, "bytes of" , len(s.m_readData), "processed")
			if(l > 0):
				s.m_readData = s.m_readData[l:]
			else:
				break;

	def tryRead_UIntData(s, input):
		try:
			return True, input.readData_UInt();
		except ChainpackDeserializationException:
			return False, -1

	def processReadData(s, read_data: bytes) -> int:
		log("processReadData data len:", len(read_data), ":", read_data)
		initial_len = len(read_data)
		input = ChainPackProtocol(read_data)
		ok, chunk_len = s.tryRead_UIntData(input);
		if not ok:
			return 0;
		received_len = len(input)
		ok, protocol_version = s.tryRead_UIntData(input);
		if not ok:
			return 0;
		if protocol_version != s.PROTOCOL_VERSION:
			raise Exception("Unsupported protocol version");
		if(chunk_len > received_len):
			return 0;
		msg: RpcValue = input.read()
		s.onMessageReceived(msg);
		return initial_len - len(input);

	def sendResponse(s, request_id: int, result: RpcValue):
		resp = RpcResponse()
		resp.setId(request_id);
		resp.setResult(result);
		info("sending response:", resp)
		s.sendMessage(resp._value);

	def sendRequest(s, method: str, params: RpcValue):
		msg = RpcRequest()
		msg.setMethod(method);
		msg.setParams(params);
		info("sending request:", method)
		s.sendMessage(msg._value);
	"""
	def sendNotify(s, method: str, result: RpcValue):
		ntf = RpcNotify()
		ntf.setMethod(method);
		ntf.setParams(result);
		info("sending notify:", method)
		s.sendMessage(ntf._value);
	"""

	"""
	def enqueueDataToSend(chunk_to_enqueue: Chunk):
		if(len(chunk_to_enqueue))
			m_chunkQueue.append(chunk_to_enqueue[:]));
		if(!isOpen()):
			logger.critical("write data error, socket is not open!")
			return
		flushNoBlock();
		writeQueue();

	def writeQueue(s):
		if(!len(s.m_chunkQueue)):
			return;
		log("writePendingData(), HI prio queue len: %s" % len(s.m_chunkQueue))
		#//static int hi_cnt = 0;
		chunk: Chunk = s.m_chunkQueue[0];

		if s.m_headChunkBytesWrittenSoFar == 0:
			protocol_version_data = ChainPackProtocol(RpcValue(PROTOCOL_VERSION, Type.UInt))
			packet_len_data = ChainPackProtocol(RpcValue(len(protocol_version_data) + len(chunk), Type.UInt))
			written_len = s.writeBytes(packet_len_data, len(packet_len_data));
			if(written_len < 0):
				raise Exception("Write socket error!");
			if written_len < len(packet_len_data):
				raise Exception("Design error! Chunk length shall be always written at once to the socket");
			
			l = s.writeBytes(protocol_version_data, len(protocol_version_data));
			if(l < 0):
				raise Exception("Write socket error!");
			if(l != 1)
				raise Exception("Design error! Protocol version shall be always written at once to the socket");

		l = s.writeBytes(chunk[s.m_headChunkBytesWrittenSoFar:])
		if(l < 0):
			raise Exception("Write socket error!");
		if(l == 0):
			raise Exception("Design error! At least 1 byte of data shall be always written to the socket");

		log("writeQueue - data len:", len(chunk),  "start index:", s.m_headChunkBytesWrittenSoFar, "bytes written:", l, "remaining:", (len(chunk) - s.m_headChunkBytesWrittenSoFar - l);
		s.m_headChunkBytesWrittenSoFar += l;
		if(s.m_headChunkBytesWrittenSoFar == len(chunk)):
			s.m_headChunkBytesWrittenSoFar = 0;
			s.m_chunkQueue.pop_front();
	"""
